edgelet/edgelet-docker/src/runtime.rs

```diff
@@ -327,46 +432,138 @@ impl ModuleRuntime for DockerModuleRuntime {
-                self.client
-                    .container_api()
-                    .container_create(create_options, module.name())
-                    .then(|result| match result {
-                        Ok(_) => Ok(module),
-                        Err(err) => Err(Error::from_docker_error(
-                            err,
-                            ErrorKind::RuntimeOperation(RuntimeOperation::CreateModule(
-                                module.name().to_string(),
-                            )),
-                        )),
+        let image_with_tag = module.config().image().to_string();
+        let digest_from_manifest = module.config().digest().map(&str::to_owned);
+
+        // check if the serveraddress exists & check if it exists in notary_registries
+        let registry_auth = module.config().auth();
+        let (registry_hostname, registry_username, registry_password) = match registry_auth {
+            Some(a) => (a.serveraddress(), a.username(), a.password()),
+            None => (None, None, None),
+        };
+        let hostname = registry_hostname.unwrap_or_default();
+        let username = registry_username.unwrap_or_default();
+        let password = registry_password.unwrap_or_default();
+
+        let image_by_notary = if !hostname.is_empty()
+            && !username.is_empty()
+            && !password.is_empty()
+        {
+            let config_path = self.notary_registries.get(hostname);
+            if let Some(c) = config_path {
+                info!("{} is enabled for notary content trust", hostname);
+                let config_path_buf = c;
+                let config_path_string_result =
+                    config_path_buf.clone().into_os_string().into_string();
+                let config_path_str;
+                match config_path_string_result {
+                    Ok(c) => config_path_str = c,
+                    Err(_) => config_path_str = "".to_string(),
+                }
+                let mut image_with_tag_parts = image_with_tag.split(':');
+                let gun = image_with_tag_parts
+                    .next()
+                    .expect("split always returns atleast one element")
+                    .to_owned();
+                let gun: Arc<str> = gun.into();
+                let tag = image_with_tag_parts.next().unwrap_or("latest").to_owned();
+                let notary_auth = format!("{}:{}", username, password);
+                let notary_auth = base64::encode(&notary_auth);
+                let lock = self.notary_lock.clone();
+                let mutex = MutexFuture(lock);
+                future::Either::A(
+                        mutex
+                            .and_then({
+                                let gun = gun.clone();
+                                move |lock| {
+                                    notary::notary_lookup(
```

Comment:
The point of making `notary_lock` hold a `BTreeMap` and populate that map in `image_pull` was to skip calling `notary_lookup` if the map already contains an entry.